cmd/daemon/main.go

```diff
@@ -417,6 +422,66 @@ func loadConfig(cfg *Config) error {
 	return nil
 }
 
+func (cfg *Config) ConfigPath() string {
+	return filepath.Join(cfg.ConfigDir, "config.yml")
+}
+
+func (cfg *Config) StatePath() string {
+	return filepath.Join(cfg.ConfigDir, "state.json")
+}
+
+func (cfg *Config) CredentialsPath() string {
+	return filepath.Join(cfg.ConfigDir, "credentials.json")
+}
+
+type AppState struct {
+	DeviceId    string                `json:"device_id"`
+	Credentials storedCredentialsFile `json:"credentials"`
+}
+
+func (app *App) readAppState() error {
+	// Read app state saved in a previous run.
+	app.state = &AppState{}
+	if content, err := os.ReadFile(app.cfg.StatePath()); err == nil {
+		if err := json.Unmarshal(content, &app.state); err != nil {
+			return fmt.Errorf("failed unmarshalling state file: %w", err)
+		}
+		log.Debugf("app state loaded")
+	} else {
+		log.Debugf("no app state found")
+	}
+
+	// Read credentials (old configuration, in credentials.json).
+	if app.state.Credentials.Username == "" {
+		if content, err := os.ReadFile(app.cfg.CredentialsPath()); err == nil {
+			if err := json.Unmarshal(content, &app.state.Credentials); err != nil {
+				return fmt.Errorf("failed unmarshalling stored credentials file: %w", err)
+			}
+
+			log.Debugf("stored credentials found for %s", app.state.Credentials.Username)
+		} else {
+			log.Debugf("stored credentials not found")
+		}
+	}
+
+	return nil
+}
+
+func (app *App) writeAppState() error {
+	err := os.MkdirAll(app.cfg.ConfigDir, 0o777)
```

Comment:
Perhaps `744` or even `700` is better?

Reply:
Yeah 700 seems like a good idea. I checked my ~/.config and it varies a lot per application whether they allow read access for groups and others but since the config file can store credentials it seems like a good idea to disallow read access.